### DISCUSSION TITLE
Paymaster Flow Issue with "zksync-ethers" Upgrade

### DISCUSSION THREAD

**Discussion by gauravl-tevaeralabs** [zkSync-Community-Hub/zksync-developers#530] [SDKs] [answered]
We recently upgraded our project from using "zksync-web3" SDK to "zksync-ethers", as "zksync-web3" is deprecated. Additionally, we updated the "ethers" package to the latest version 6.x. However, after the upgrade, we encountered an issue with the paymaster flow.

Previously, with zksync-web3, the paymaster flow was functioning correctly. However, after the upgrade, the paymaster no longer works as expected. Instead of displaying the paymaster approval dialogue, it redirects the user to the normal flow where transaction approval is done with ETH.

We suspect that there might be some syntactical changes introduced in the new SDK, but the documentation hasn't been updated to reflect these changes. We're seeking assistance to resolve this issue.

**Code Snippet:**
```
const contract = new Contract(
  contractAddress,
  require("./abi.json"),
  connectedWallet
);

// create paymaster params for general flow
const paymasterParams = utils.getPaymasterParams(paymasterContractAddress, {
  type: "General",
  innerInput: new Uint8Array(),
});

// call register with paymaster params
const txn = await contract.register({
  customData: {
    paymasterParams,
    gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
  },
});
const { hash } = txn;
```

Any guidance or insights into resolving this issue would be greatly appreciated. Thank you!

**Comment by dutterbutter** [comment]
Hey @gauravl-tevaeralabs 👋 

The syntax here is correct. This sentence " Instead of displaying the paymaster approval dialogue, it redirects the user to the normal flow where transaction approval is done with ETH." makes me think its likely this line `gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,`. 

Just to test can you do something like `gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT + 80000n` and see if that changes the behaviour here.

**Reply by gauravl-tevaeralabs** [reply]
Hi @dutterbutter, I attempted to increase the default gas per pubdata limit by 80k as per your suggestion, but unfortunately, it didn't resolve the issue. I don't believe it's related to the gas limit, as I'm utilizing a GeneralFlow, which ideally should function even if the user lacks gas fees, as we're funding the gas based on an arbitrary condition set by the paymaster

**Reply by bxpana** [reply]
So far everything looks all good. Are you guys able to share a link to the repo or invite us to it?

**Reply by gauravl-tevaeralabs** [reply]
Hi @bxpana,
I've created a demo to demonstrate an issue with "zksync-ethers" where the same code and paymaster work with the "zksync-web3" SDK. Please clone this public GitHub repo and follow the README file. Let me know if you need any additional information or clarification. It would be really helpful if you could point out any mistakes I might be making. Thanks!

Github Repo: https://github.com/gauravl-tevaeralabs/zksync-paymaster-demo

**Reply by tevanrobin** [reply]
Any update here team? @dutterbutter @bxpana

**Reply by tevanrobin** [reply]
@dutterbutter @bxpana - any update team?

**Reply by 0xsisyfos** [reply, marked as answer]
Sorry fro the late answer.

 `Wallet.createRandom()` does not function the same way in v6 as in v5. So random wallet should be created like in the example below.
Example:
```
async function registerQuestUsingPaymaster() {
  // create zksync provider
  const provider = new Provider(zksyncRpcUrl);
  // generate random wallet
  const wallet = new Wallet(Wallet.createRandom().privateKey, provider, null);

  // register quest with newly generated wallet
  await registerQuest(wallet);
}
```

**Reply by gauravl-tevaeralabs** [reply]
Thanks for your reply, @0xsisyfos. The suggested syntax is working. We're actually using the window.ethereum provider with a signer, so we'll look into our code. Most likely, there might be some syntax changes needed as zksync-ethers seems to have no issues. Thanks again!
